src/transformers/models/yoso/modeling_yoso.py

```diff
@@ -0,0 +1,1690 @@
+
+        if not self.use_expectation:
+            loaded = load_cuda_kernels()
+
+            if not loaded:
+                raise ValueError(
+                    """Failed to compile CUDA kernels. Ensure that the correct versions of PyTorch and cudatoolkit
+                    are installed and try again. Alternatively, avoid loading CUDA kernels by using YOSO Expectation by
+                    setting use_expectation=True."""
+                )
+
+    def transpose_for_scores(self, layer):
+        new_layer_shape = layer.size()[:-1] + (self.num_attention_heads, self.attention_head_size)
+        layer = layer.view(*new_layer_shape)
+        return layer.permute(0, 2, 1, 3)
+
+    def forward(
+        self,
+        hidden_states,
+        attention_mask=None,
+        head_mask=None,
+        encoder_hidden_states=None,
+        encoder_attention_mask=None,
+        past_key_value=None,
+        output_attentions=False,
+    ):
+        mixed_query_layer = self.query(hidden_states)
+
+        key_layer = self.transpose_for_scores(self.key(hidden_states))
+        value_layer = self.transpose_for_scores(self.value(hidden_states))
+        query_layer = self.transpose_for_scores(mixed_query_layer)
+
+        if self.use_conv:
+            conv_value_layer = self.conv(value_layer * attention_mask[:, None, :, None])
+
+        batch_size, num_heads, seq_len, head_dim = query_layer.size()
+
+        query_layer = query_layer.reshape(batch_size * num_heads, seq_len, head_dim)
+        key_layer = key_layer.reshape(batch_size * num_heads, seq_len, head_dim)
+        value_layer = value_layer.reshape(batch_size * num_heads, seq_len, head_dim)
+
+        # revert changes made by get_extended_attention_mask
+        attention_mask = 1.0 + attention_mask / 10000.0
+        attention_mask = attention_mask.squeeze().repeat(1, num_heads, 1).reshape(batch_size * num_heads, seq_len)
+
+        # The CUDA kernels are most efficient with inputs whose size is a multiple of a GPU's warp size (32). Inputs
+        # smaller than this are padded with zeros.
+        gpu_warp_size = 32
+
+        if (not self.use_expectation) and head_dim < gpu_warp_size:
+            query_layer = torch.cat(
+                [
+                    query_layer,
+                    torch.zeros(batch_size * num_heads, seq_len, gpu_warp_size - head_dim, device=query_layer.device),
+                ],
+                dim=-1,
+            )
+            key_layer = torch.cat(
+                [
+                    key_layer,
+                    torch.zeros(batch_size * num_heads, seq_len, gpu_warp_size - head_dim, device=key_layer.device),
+                ],
+                dim=-1,
+            )
+            value_layer = torch.cat(
+                [
+                    value_layer,
+                    torch.zeros(batch_size * num_heads, seq_len, gpu_warp_size - head_dim, device=value_layer.device),
+                ],
+                dim=-1,
+            )
```

Comment:
This could be made simpler to read by isolating `batch_size * num_heads, seq_len, gpu_warp_size - head_dim` in a variable